### PR TITLE
More speedups/fixes to creating batched events

### PR DESCRIPTION
**Pull request by H-Shay**
These changes come from a discussion with Erik and are follow-ups to the batching persisting event work, speeding it up/making it more efficient. Best reviewable commit-by-commit, and I can split these out into separate PRs if others think it necessary.

### CODE REVIEW

File: synapse/event_auth.py
Comment (clokep) [inline synapse/event_auth.py:181]: It might make more sense to keep the difference instead of calculating it twice:

```suggestion
        auth_event_ids = event.auth_event_ids()
        auth_events = dict(batched_auth_events)
        needed_auth_event_ids = set(event.auth_event_ids()) - batched_auth_events.keys()
        if needed_auth_event_ids:
            auth_events.update(
                await store.get_events(
                    needed_auth_event_ids,
                    redact_behaviour=EventRedactBehaviour.as_is,
                    allow_rejected=True,
                )
            )
```

File: synapse/handlers/event_auth.py
Comment (clokep) [inline synapse/handlers/event_auth.py:79]: Same here:

```suggestion
            auth_events_by_id = dict(batched_auth_events)
            needed_auth_event_ids = set(auth_event_ids) - set(batched_auth_events)
            if needed_auth_event_ids:
                logger.info("fetching %s", needed_auth_event_ids)
                auth_events_by_id.update(
                    await self._store.get_events(needed_auth_event_ids)
```

(Are we sure we want that log at `info`? Is this sometimes a big list?)

File: synapse/event_auth.py
Comment (H-Shay) [reply synapse/event_auth.py]: thanks, that makes sense!

File: synapse/handlers/event_auth.py
Comment (H-Shay) [reply synapse/handlers/event_auth.py]: Now that you mention it I think the logging is not necessary in general, I've dropped it.

File: synapse/event_auth.py
Comment (clokep) [inline synapse/event_auth.py:172]: I keep wondering why we're doing this:

```suggestion
        # Copy the batched auth events to avoid mutating them.
        auth_events = dict(batched_auth_events)
```

File: synapse/events/snapshot.py
Comment (clokep) [inline synapse/events/snapshot.py:138]: This needs a comment in the docstring.

File: synapse/handlers/room.py
Comment (clokep) [inline synapse/handlers/room.py:1126]: I don't understand why we need to take a copy here?

File: synapse/events/snapshot.py
Comment (clokep) [reply synapse/events/snapshot.py:138]: I'm failing to see where this ever gets set when creating an `EventContext`.

File: synapse/handlers/room.py
Comment (H-Shay) [reply synapse/handlers/room.py:1126]: When Erik suggested it I think he said something about not mutating the state map but upon your question I looked closer and a: I don't_ see where it's mutated in the functions it's passed to, and b: it's immediately mutated right below so now I am also confused and will clarify it with him.

Comment (clokep) [reply synapse/handlers/room.py:1126]:
```suggestion
                # Take a copy to ensure each event gets a unique copy of
                # state_map since it is modified below.
                state_map=dict(state_map),
```

Comment (clokep) [reply synapse/handlers/room.py:1126]: So is this fixing an actual bug then?

Comment (H-Shay) [reply synapse/handlers/room.py:1126]: Yes, I think so.

Comment (clokep) [reply synapse/handlers/room.py:1126]: Should it have a test?!

File: synapse/events/snapshot.py
Comment (H-Shay) [reply synapse/events/snapshot.py:138]: I think this is to bring EventContext into alignment with UnpersistedEventContext, from which it inherits and which has this field. There's also a separate fix I am working on that will use this field.

File: synapse/handlers/room.py
Comment (H-Shay) [reply synapse/handlers/room.py:1126]: I am judging by the exclamation point that the answer to your question is yes. I hadn't considered that, I will add one.

Comment (clokep) [reply synapse/handlers/room.py:1126]: I'm not sure if it is or not, but it seems like a thing we could mistakenly reintroduce by refactoring at some point in the future?

File: synapse/events/snapshot.py
Comment (clokep) [reply synapse/events/snapshot.py:138]: (Note necessarily for this PR!)

File: synapse/handlers/room.py
Comment (H-Shay) [reply synapse/handlers/room.py:1126]: I guess it comes down to whether you think the comment is enough of a guard to avoid that - I can see both perspectives, but since the test should be pretty easy to implement I can just add it.

Comment (H-Shay) [reply synapse/handlers/room.py:1126]: Actually now that I've tried to test this I am not sure it's possible - `create_event`, the result of which I am trying to test is a function that is internal to `_send_events_for_new_room`.  `_send_events_for_new_room` consumes the results of `create_event` (an `UnpersistedEventContext` and a new event), converts the `UnpersistedEventContext` into an `EventContext` which then gets sent to the database with the new event, but `EventContexts` aren't stored, so I have no way to access the rather ephemeral `state_map` in question. Note that this bug doesn't affect the actual state or state group for the event. Given this I honestly think that just adding the comment should be fine, but I leave this up to your expertise.

Comment (clokep) [reply synapse/handlers/room.py:1126]: If it doesn't get stored in the database then I agree it doesn't seem worth testing.

File: changelog.d/15195.misc
Comment (clokep) [inline changelog.d/15195.misc:1]:
```suggestion
Improve performance of creating and authenticating events.
```

File: synapse/events/snapshot.py
Comment (clokep) [reply synapse/events/snapshot.py:138]: Actually -- should this be part of `UnpersistedEventContextBase` instead?

> There's also a separate fix I am working on that will use this field.

I think we should derfer the changes to this file until we're ready to use them, otherwise we might end up with dead code. (Sorry for the change of heart.)

**Review by clokep — CHANGES_REQUESTED** [review]

File: synapse/events/snapshot.py
Comment (H-Shay) [reply synapse/events/snapshot.py]: No problem!

**Review by clokep — APPROVED** [review]
LGTM if CI is happy.

**Comment by H-Shay** [comment]
Merged #15195 into main, thanks for the review.

**Pull request merged**
